arangod/Pregel/Algos/AIR/GraphFormat.cpp

```diff
@@ -0,0 +1,209 @@
+#include "Pregel/Algos/AIR/Greenspun/Interpreter.h"
+
+#include <velocypack/Collection.h>
+#include <velocypack/velocypack-aliases.h>
+
+namespace arangodb {
+namespace pregel {
+namespace algos {
+namespace accumulators {
+
+// Graph Format
+GraphFormat::GraphFormat(application_features::ApplicationServer& server,
+                         std::string const& resultField,
+                         AccumulatorsDeclaration const& globalAccumulatorDeclarations,
+                         AccumulatorsDeclaration const& vertexAccumulatorDeclarations,
+                         CustomAccumulatorDefinitions customDefinitions,
+                         DataAccessDefinition const& dataAccess)
+    : graph_format(server),
+      _resultField(resultField),
+      _globalAccumulatorDeclarations(globalAccumulatorDeclarations),
+      _vertexAccumulatorDeclarations(vertexAccumulatorDeclarations),
+      _customDefinitions(std::move(customDefinitions)),
+      _dataAccess(dataAccess) {}
+
+size_t GraphFormat::estimatedVertexSize() const { return sizeof(vertex_type); }
+size_t GraphFormat::estimatedEdgeSize() const { return sizeof(edge_type); }
+
+void filterDocumentData(VPackBuilder& finalBuilder, VPackSlice const& arraySlice,
+                        arangodb::velocypack::Slice& document) {
+  for (auto&& key : VPackArrayIterator(arraySlice)) {
+    if (key.isString()) {
+      VPackBuilder tmp;
+      VPackBuilder innerBuilder;
+      innerBuilder.openObject();
+      innerBuilder.add(key.copyString(), document.get(key.stringRef()));
+      innerBuilder.close();
+
+      if (!finalBuilder.slice().isObject()) {
+        finalBuilder.openObject();
+        finalBuilder.close();
+      }
+      VPackCollection::merge(tmp, finalBuilder.slice(), innerBuilder.slice(), true, false);
+      finalBuilder.clear();
+      finalBuilder.add(tmp.slice());
+    } else if (key.isArray()) {
+      std::vector<VPackStringRef> path;
+      size_t pathLength = key.length();
+      size_t iterationStep = 0;
+
+      VPackBuilder innerArrayBuilder;
+      innerArrayBuilder.openObject();  // open outer object
+
+      for (auto&& innerKey : VPackArrayIterator(key)) {
+        // TODO: optimize in the future - right now complicated to build a fresh object via key-path strings
+
+        if (!innerKey.isString()) {
+          TRI_ASSERT(false);  // TODO: Add type checking in deserializer
+        }
+        // build up path - will change in every iteration step
+        path.emplace_back(innerKey.stringRef());
+
+        if (iterationStep < (pathLength - 1)) {
+          innerArrayBuilder.add(innerKey);  //, VPackValue(VPackValueType::Object));
+          innerArrayBuilder.openObject();
+        } else {
+          innerArrayBuilder.add(innerKey.copyString(), document.get(path));
+        }
+
+        // get slice of each document depth
+        iterationStep++;
+      }
+
+      // now close all inner opened objects
+      for (size_t step = 0; step < (pathLength - 1); step++) {
```

Comment:
If `pathLength` is 0, then this loop will be very unhappy.

Reply:
@hkernbach what is the correct behavior if `pathLength == 0`?

Reply:
The deserializer now ensures that pathLength is `> 0`